framework/src/modules/bft/api.ts

```diff
@@ -12,45 +12,253 @@
+		return paramsStore.has(intToBuffer(height, 4, BIG_ENDIAN));
+	}
+
+	public async getBFTParameters(
+		context: ImmutableAPIContext,
+		height: number,
+	): Promise<BFTParameters> {
+		const paramsStore = context.getStore(this.moduleID, STORE_PREFIX_BFT_PARAMETERS);
+		return getBFTParameters(paramsStore, height);
+	}
+
+	public async getBFTHeights(context: ImmutableAPIContext): Promise<BFTHeights> {
+		const votesStore = context.getStore(this.moduleID, STORE_PREFIX_BFT_VOTES);
+		const bftVotes = await votesStore.getWithSchema<BFTVotes>(EMPTY_KEY, bftVotesSchema);
 		return {
-			maxHeightPrecommited: 0,
-			maxHeightPrevoted: 0,
+			maxHeightPrevoted: bftVotes.maxHeightPrevoted,
+			maxHeightPrecommited: bftVotes.maxHeightPrecommited,
+			maxHeightCertified: bftVotes.maxHeightCertified,
+		};
+	}
+
+	public async impliesMaximalPrevotes(
+		context: ImmutableAPIContext,
+		header: BlockHeader,
+	): Promise<boolean> {
+		const votesStore = context.getStore(this.moduleID, STORE_PREFIX_BFT_VOTES);
+		const bftVotes = await votesStore.getWithSchema<BFTVotes>(EMPTY_KEY, bftVotesSchema);
+		const [lastHeader] = bftVotes.blockBFTInfos;
+		if (header.height !== lastHeader.height + 1) {
+			return false;
+		}
+		const previousHeight = header.maxHeightGenerated;
+
+		// the block does not imply any prevotes
+		if (previousHeight >= header.height) {
+			return false;
+		}
+
+		// there is no block info stored for previousHeight and header implies the maximal number of prevotes
+		const offset = lastHeader.height - previousHeight;
+		if (offset >= bftVotes.blockBFTInfos.length) {
+			return true;
+		}
+		// block at previousHeight is generated by a different delegate and header doesn't
+		// imply maximal number of prevotes
+		if (!bftVotes.blockBFTInfos[offset].generatorAddress.equals(header.generatorAddress)) {
+			return false;
+		}
+		return true;
+	}
+
+	public async getNextHeightBFTParameters(
+		context: ImmutableAPIContext,
+		height: number,
+	): Promise<number> {
+		const paramsStore = context.getStore(this.moduleID, STORE_PREFIX_BFT_PARAMETERS);
+		const start = intToBuffer(height + 1, 4, BIG_ENDIAN);
+		const end = intToBuffer(MAX_UINT32, 4, BIG_ENDIAN);
+		const results = await paramsStore.iterate({
+			limit: 1,
+			start,
+			end,
+		});
+		if (results.length !== 1) {
+			throw new BFTParameterNotFoundError();
+		}
+		const [result] = results;
+		return result.key.readUInt32BE(0);
+	}
+
+	public async setBFTParameters(
+		context: APIContext,
+		precommitThreshold: bigint,
+		certificateThreshold: bigint,
+		validators: Validator[],
+	): Promise<void> {
+		let aggregateBFTWeight = BigInt(0);
+		for (const validator of validators) {
+			if (validator.bftWeight <= 0) {
+				throw new Error('Invalid BFT weight. BFT weight must be a positive integer.');
+			}
+			aggregateBFTWeight += validator.bftWeight;
+		}
+		if (
+			aggregateBFTWeight / BigInt(3) + BigInt(1) > precommitThreshold ||
```

Comment:
Just to confirm, we do integer division here so the left side is equivalent to
`floor(1/3*aggregateBFTWeight)+1`, right?

Reply:
yes, int calculation is `floor` as default so left one is equivalent to the `floor(1/3*aggregateBFTWeight)+1`

Reply:
```
> BigInt(103) / BigInt(3) + BigInt(1)
35n
```

Reply:
All good then :-)